server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -137,12 +138,15 @@ private static BulkItemResultHolder executeIndexRequest(final IndexRequest index
                                                             final IndexShard primary,
                                                             final MappingUpdatePerformer mappingUpdater) throws Exception {
         Engine.IndexResult indexResult = executeIndexRequestOnPrimary(indexRequest, primary, mappingUpdater);
-        if (indexResult.hasFailure()) {
-            return new BulkItemResultHolder(null, indexResult, bulkItemRequest);
-        } else {
-            IndexResponse response = new IndexResponse(primary.shardId(), indexRequest.type(), indexRequest.id(),
+        switch (indexResult.getResultType()) {
```

Comment:
what was wrong with a boolean here? What does this result type buy us?

Reply:
yeah, that's a good question. I was doubting on how to do this and ended with with the enum but I'm not 100% happy with it either. I wanted `IndexShard.applyIndexOperationOnPrimary` to always return without the extra callbacks. That method can end with success, failure or a required mapping change (I didn't want to make this a failure). The problem with the booleans is you need to coordinate them - make sure that there is no failure and a required mapping update. I hope this clarifies things. I'm open to suggestions.

Reply:
I am ok with it. stuff like this is always tricky